src/state_machine.rs

```diff
@@ -0,0 +1,1732 @@
+}
+
+#[derive(Default, Clone, Debug)]
+/// Map of signer_id and key_id to the relevant ecdsa public keys
+pub struct PublicKeys {
+    /// signer_id -> public key
+    pub signers: HashMap<u32, ecdsa::PublicKey>,
+    /// key_id -> public key
+    pub key_ids: HashMap<u32, ecdsa::PublicKey>,
+}
+
+/// State machine for a simple FROST coordinator
+pub mod coordinator {
+    use hashbrown::HashSet;
+    use p256k1::{point::Point, scalar::Scalar};
+    use std::collections::BTreeMap;
+    use tracing::info;
+
+    use crate::{
+        common::{MerkleRoot, PolyCommitment, PublicNonce, Signature, SignatureShare},
+        compute,
+        errors::AggregatorError,
+        net::{
+            DkgBegin, DkgPublicShares, Message, NonceRequest, NonceResponse, Packet, Signable,
+            SignatureShareRequest,
+        },
+        state_machine::{OperationResult, StateMachine},
+        taproot::SchnorrProof,
+        traits::Aggregator as AggregatorTrait,
+    };
+
+    #[derive(Debug, PartialEq)]
+    /// Coordinator states
+    pub enum State {
+        /// The coordinator is idle
+        Idle,
+        /// The coordinator is distributing public shares
+        DkgPublicDistribute,
+        /// The coordinator is gathering public shares
+        DkgPublicGather,
+        /// The coordinator is distributing private shares
+        DkgPrivateDistribute,
+        /// The coordinator is gathering DKG End messages
+        DkgEndGather,
+        /// The coordinator is requesting nonces
+        NonceRequest(bool, Option<MerkleRoot>),
+        /// The coordinator is gathering nonces
+        NonceGather(bool, Option<MerkleRoot>),
+        /// The coordinator is requesting signature shares
+        SigShareRequest(bool, Option<MerkleRoot>),
+        /// The coordinator is gathering signature shares
+        SigShareGather(bool, Option<MerkleRoot>),
+    }
+
+    #[derive(thiserror::Error, Debug)]
+    /// The error type for the coordinator
+    pub enum Error {
+        /// A bad state change was made
+        #[error("Bad State Change: {0}")]
+        BadStateChange(String),
+        /// A bad dkg_id in received message
+        #[error("Bad dkg_id: got {0} expected {1}")]
+        BadDkgId(u64, u64),
+        /// A bad sign_id in received message
+        #[error("Bad sign_id: got {0} expected {1}")]
+        BadSignId(u64, u64),
+        /// A bad sign_iter_id in received message
+        #[error("Bad sign_iter_id: got {0} expected {1}")]
+        BadSignIterId(u64, u64),
+        /// SignatureAggregator error
+        #[error("Aggregator: {0}")]
+        Aggregator(AggregatorError),
+        /// Schnorr proof failed to verify
+        #[error("Schnorr Proof failed to verify")]
+        SchnorrProofFailed,
+    }
+
+    impl From<AggregatorError> for Error {
+        fn from(err: AggregatorError) -> Self {
+            Error::Aggregator(err)
+        }
+    }
+
+    /// Coordinatable trait for handling the coordination of DKG and sign messages
```

Comment:
Might make sense to have coordinatable and Coordinator live outside the state_machine.rs file in something like corodinator.rs or have a subdirectory with the following: frost/coordinator.rs. and frost/state_machine.rs

Reply:
I'm still trying to decide how to organize all of this.  Before I stopped working on Friday, my plan was to have only `StateMachine` and `OperationResult` live in `state_machine.rs`, then have a `coordinator.rs` with `Coordinatable`, and sub modules for `frost`, `fire`, and `roast`, and a `signer.rs` with `SigningRound`.

I hadn't decided whether to put the `coordinator` and `signer` mods into `state_machine` or leave them as top-level.  What are your thoughts about this org?

Reply:
I'm gonna go ahead and merge this, tag it as `4.0.0rc1`, then merge the `stacks-signer` PR.  I'll wait to release the `4.0.0` crate until we've decided on how to organize the modules.